### DISCUSSION TITLE
C# 12 - Definitely supported features

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#7493] [General] [answered]
I know that C# 12 is not yet released and I don't care what date it will eventually be released, but if I wanted to take it for granted, that when it _is_ released it will _definitely_ contain certain features, then what is that definite list of features? 

Clearly "interceptors" are not something we can assume will definitely be available when C# 12 gets released, but what about the other listed features [here ](https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md#c-120)and [here](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12) and [here](https://devblogs.microsoft.com/dotnet/check-out-csharp-12-preview/)?

If I were to "promise" some team of developers that when we get C#12 we will definitely be able to leverage say primary constructors or collection expressions or any of the other listed non-experimental features, would that be safe to do?

**Comment by CyrusNajmabadi** [comment]
> that when it is released it will definitely contain certain features, then what is that definite list of features?

We cannot give definites on future releases.  All information is subject to change.  

> If I were to "promise" some team of developers that when we get C#12 we will definitely be able to leverage say primary constructors or collection expressions or any of the other listed non-experimental features, would that be safe to do?

No.  Until the release actually happens, all features are subject to change and potential withdrawal.  

--

However, if you want to instead give reasonable advise of what is most likely, then we can supply you with that information.

**Reply by CyrusNajmabadi** [reply, marked as answer]
For the tentative list, see: https://github.com/dotnet/roslyn/discussions/69074

**Reply by The-Futurist** [reply]
Many thanks, that's helpful.

**Reply by jnm2** [reply]
As for _when_, the C# language follows the release cadence of .NET, and .NET has been releasing a new major version each year in the second week of November, if I remember right.
